src/draw/nxp/g2d/lv_g2d_buf_map.c

```diff
@@ -0,0 +1,251 @@
+
+    if(item == NULL) {
+        return;
+    }
+    else if(list == NULL && item->key == key) {
+        /* No collision chain, just remove item. */
+        table->items[index] = NULL;
+        _map_free_item(item);
+        table->count--;
+        return;
+    }
+    else if(list != NULL) {
+        /* Collision chain exists. */
+        for(uint32_t i = 0; i < lv_array_size(list); i++) {
+            item = (lv_map_item_t *)lv_array_at(list, i);
+            if(item->key == key) {
+                lv_array_remove(list, i);
+                return;
+            }
+        }
+    }
+}
+
+void g2d_print_table(void)
+{
+    LV_LOG("\nHash Table\n-------------------\n");
+
+    for(int i = 0; i < table->size; i++) {
+        if(table->items[i]) {
+            LV_LOG("Index:%d, Key:%p, Value:%p\n", i, table->items[i] -> key, table->items[i]->value);
+            if(table->overflow_list[i]) {
+                for(int j = 0 ; j < lv_array_size(table->overflow_list[i]); j++) {
+                    lv_map_item_t * item = (lv_map_item_t *)lv_array_at(table->overflow_list[i], j);
+                    LV_LOG("Index:%d, Key:%p, Value:%p\n", i, item -> key, item->value);
+                }
+
+            }
+        }
+    }
+
+    LV_LOG("-------------------\n\n");
+}
+
+/**********************
+ *   STATIC FUNCTIONS
+ **********************/
+
+static unsigned long _map_hash_function(void * ptr)
+{
+    unsigned long i = 0;
+    char str[64];
+    G2D_ASSERT_MSG(ptr, "Key is null.");
+    sprintf(str, "%p", ptr);
+
+    for(int j = 0; str[j]; j++)
+        i += str[j];
+
+    return i % HASH_TABLE_SIZE;
+}
+
+static void _handle_collision(unsigned long index, lv_map_item_t * item)
+{
+    if(table->overflow_list[index] == NULL) {
+        /* Create the list. */
+        lv_array_t * list = (lv_array_t *) lv_malloc(sizeof(lv_array_t));;
+        lv_array_init(list, LV_ARRAY_DEFAULT_CAPACITY, sizeof(lv_map_item_t));
+        lv_array_push_back(list, item);
+        table->overflow_list[index] = list;
+        return;
+    }
+    else {
+        lv_array_t * list = (lv_array_t *)table->overflow_list[index];
+        for(int i = 0; i < lv_array_size(list); i++) {
+            lv_map_item_t * it = (lv_map_item_t *)lv_array_at(list, i);
+            if(it->key == item->key) {
+                /* Key exists, update value. */
+                it->value = item->value;
+                return;
+            }
+        }
+        /* Insert to the list. */
+        lv_array_push_back(table->overflow_list[index], item);
+        return;
+    }
+}
+
+static lv_map_item_t * _map_create_item(void * key, struct g2d_buf * value)
+{
+    lv_map_item_t * item = (lv_map_item_t *) lv_malloc(sizeof(lv_map_item_t));
+    G2D_ASSERT_MSG(item, "Failed to alloc item.");
+    item->key = key;
+    item->value = value;
+    return item;
+}
+
+static void _map_free_item(lv_map_item_t * item)
+{
+    /* Also free the g2d_buf. */
+    lv_free(item->key);
+    g2d_free(item->value);
```

Comment:
It seems this came from `lv_malloc`
```suggestion
    lv_free(item->value);
```

Reply:
The item indeed comes from `lv_malloc`, but the value is a `struct g2d_buf`, which comes from a `g2d_alloc`, so it needs to be freed with `g2d_free`. Here are the only places where value is allocated: [image allocation](https://github.com/nxp-mcuxpresso/lvgl/blob/ca767e33a97f10bcfcf3484dd0ce72de3707d3c9/src/draw/nxp/g2d/lv_draw_g2d_img.c#L131), [buf_malloc](https://github.com/nxp-mcuxpresso/lvgl/blob/ca767e33a97f10bcfcf3484dd0ce72de3707d3c9/src/draw/nxp/g2d/lv_draw_buf_g2d.c#L73).

Reply:
> The item indeed comes from `lv_malloc`, but the value is a `struct g2d_buf`, which comes from a `g2d_alloc`, so it needs to be freed with `g2d_free`. Here are the only places where value is allocated: [image allocation](https://github.com/nxp-mcuxpresso/lvgl/blob/ca767e33a97f10bcfcf3484dd0ce72de3707d3c9/src/draw/nxp/g2d/lv_draw_g2d_img.c#L131), [buf_malloc](https://github.com/nxp-mcuxpresso/lvgl/blob/ca767e33a97f10bcfcf3484dd0ce72de3707d3c9/src/draw/nxp/g2d/lv_draw_buf_g2d.c#L73).

Oh yeah! Thanks for clarifying.